mender-update/daemon/states.hpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_UPDATE_STATES_HPP
+#define MENDER_UPDATE_STATES_HPP
+
+#include <common/io.hpp>
+#include <common/state_machine.hpp>
+
+#include <artifact/artifact.hpp>
+
+#include <mender-update/daemon/context.hpp>
+#include <mender-update/daemon/state_events.hpp>
+
+namespace mender {
+namespace update {
+namespace daemon {
+
+using namespace std;
+
+namespace io = mender::common::io;
+namespace sm = mender::common::state_machine;
+
+namespace artifact = mender::artifact;
+
+using StateType = sm::State<Context, StateEvent>;
+
+class IdleState : virtual public StateType {
+public:
+	void OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) override;
+};
+
+class PollForDeploymentState : virtual public StateType {
+public:
+	PollForDeploymentState(events::EventLoop &loop) :
+		poll_timer_(loop) {
+	}
+	void OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) override;
+
+private:
+	events::Timer poll_timer_;
+};
+
+class SubmitInventoryState : virtual public StateType {
+public:
+	SubmitInventoryState(events::EventLoop &loop) :
+		poll_timer_(loop) {
+	}
+	void OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) override;
+
+private:
+	events::Timer poll_timer_;
+};
+
+class UpdateDownloadState : virtual public StateType {
+public:
+	void OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) override;
+
+	static void ParseArtifact(Context &ctx, sm::EventPoster<StateEvent> &poster);
+	static void FillUpdateDataFromArtifact(artifact::PayloadHeaderView &header, StateData &data);
```

Comment:
It's simply because they don't use anything from `this`, but still belong here logically (ignoring your other comment about `FillUpdateDataFromArtifact` for this dicsussion). Isn't it kind of self evident for a `static` member function?

Reply:
But they also don't use anything from the class, do they? So they could be just independent functions (or somewhere else, but that's the other comment/discussion).

Reply:
Yeah, so I'm moving `FillUpdateDataFromArtifact` as you pointed out.

The other one I thought could stay as it is still very specific to this class, even if it doesn't use `this`, it only uses the two arguments. No strong opinion on whether it should be `static`. I could remove it just so it looks more tightly integrated with the class.

Reply:
Like I wrote above, I think a comment would do.

Reply:
Handled in #1315.